### DISCUSSION TITLE
关于 jsv8

### DISCUSSION THREAD

**Discussion by vfleaking** [vfleaking/uoj#88] [General]
在编译 Markdown 时，UOJ 使用了 jsv8 来在 PHP 程序里调用一段 js 脚本。
（这里有一些历史原因。。因为很久以前 UOJ 的 Markdown 编译需要既在客户端也需要在服务端进行。。客户端的是为了预览。。所以才选择了 js。。后来放弃了这一设计，直接传给服务端编译了，所以会出现让 PHP 执行 js 的情况）

但是当前版本里面，docker 容器并不会自带 jsv8，导致博客功能无法直接使用。
这是因为 Ubuntu 20.04 里面 PHP 的 jsv8 不再能很方便地安装了。。

所以下面我来简单地贴一下当时我在给 Ubuntu 20.04 装 jsv8 时记的笔记，装好后应该就能正常使用博客功能了。

#### 第一步
https://github.com/phpv8/v8js/blob/php7/README.Linux.md
根据这里的指导编译v8。版本选择的是: 7.5.288.31

#### 第二步
把编译好后的文件传到服务器上去，然后在服务器上执行（切换到一个能看到 `out.gn` 的目录）：
```bash
mkdir -p /opt/v8/{lib,include}
cp out.gn/x64.release/lib*.so out.gn/x64.release/*_blob.bin out.gn/x64.release/icudtl.dat /opt/v8/lib
cp -R include/* /opt/v8/include
```

#### 第三步
使用 `pecl install` 安装 v8js
```bash
echo /opt/v8 | pecl install v8js
```

#### 第四步
启用 v8js
```bash
echo extension=v8js >/etc/php/7.4/mods-available/v8js.ini
ln -s /etc/php/7.4/mods-available/v8js.ini /etc/php/7.4/apache2/conf.d/20-v8js.ini
```

**Comment by SunnyORcold** [comment]
v8js 在uoj里，主要作用就是富文本。
php 支持v8js 主要分为两个方面，libv8 和 v8js。一个是系统支持，一个是php 扩展。

问题点：
        1.ubuntn 系统源 只有libv8 3.4版本
        2.阿里源 没有
        3.谷歌有但是访问不到。要梯子
谷歌： https://chromium.googlesource.com/chromium/tools/depot_tools.git

国内源：
### 下载 depot_tools 工具
git clone https://source.codeaurora.org/quic/lc/chromium/tools/depot_tools
export PATH=`pwd`/depot_tools:"$PATH"

### 下载v8。这个东西很大，下的贼慢
1. git clone https://source.codeaurora.org/quic/la/platform/external/chromium_org/v8 //下载慢，还需要自己检出分支
2.  https://github.com/v8/v8 //下载快

**Comment by qwerty472123** [comment]
一个较快的方法：

```bash
# pull 社区版
sudo docker pull universaloj/uoj-system
# 启动社区版
sudo docker run --name uoj -dit --cap-add SYS_PTRACE universaloj/uoj-system
# 复制出 libv8
sudo docker cp uoj:/opt/libv8-7.5 ./
sudo docker cp ./libv8-7.5 uoj_all:/opt

sudo docker exec -it uoj_all /bin/bash
# 以下指令在exec 中
# 安装 v8js
echo /opt/libv8-7.5 | pecl install v8js
# 启用 v8js
echo extension=v8js >/etc/php/7.4/mods-available/v8js.ini
ln -s /etc/php/7.4/mods-available/v8js.ini /etc/php/7.4/apache2/conf.d/20-v8js.ini
service apache2 restart
exit
```

**Comment by vfleaking** [comment]
前几天我交了个commit：https://github.com/vfleaking/uoj/commit/1ace7d321b3977d1e901d8ceaffba995c3f0c7f5
现在 UOJ 安装后默认会通过一个内部的 node 服务器编译 markdown 咯，不再需要 v8js 库
如果有同学想怀旧，可以往数据库中的 meta 表中加一项 key 为 `markdown`，value 为 `{"backend":"v8js"}` 的条目，就可以切换回调用 v8js 来编译的模式（当然，你需要像以前一样自己折腾好 v8js）
